### ISSUE TITLE
Switch expressions behavior unspecified when not exhaustive

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#2123]
The [patterns proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md#switch-expression) includes an expression switch.
It does not say what happens if no case matches the expression.

One example given is:
```dart
var x = switch (n) {
  case 1 => obj;
  case 2 => another;
}.someMethod();
```

This looks like it should be invalid because the switch is non-exhaustive, and you can't call `.someMethod` on nothing.

Either the switch throws on a non-match, in which case the example is valid, or it evaluates to a value (likely `null`), and then the example is invalid. (Or it's a compile-time error and you have to provide a `default:` case.)

Throwing on being non-exhaustive is subtly different from switch statement behavior, and is a hidden run-time error, so probably not a good match for the current Dart.

**Comment by mateusfccp** [comment]
> Or it's a compile-time error and you have to provide a `default:` case.

Although not stated, I think this is the idea. It should be a compile-time error to non-exhausively use a switch expression.

**Comment by Jetz72** [comment]
I think the default `default` returning `null` would be ideal if I encountered a case like the given example. I could either throw in a `?.` to skip the method call for outlying cases, or a `!.` to assert that it is actually exhaustive in practice, or just add a `default` clause if that seems like the most appropriate solution. Under this behavior, a switch-expression without a default clause can be statically be considered nullable, so returning it from a non-nullable function or storing it in a non-nullable variable will require the same resolution.

Suppose one drawback would be that it could end up hiding an error if one or more cases is also nullable. Perhaps a warning would suffice for that?

**Comment by Levi-Lesches** [comment]
> Suppose one drawback would be that it could end up hiding an error if one or more cases is also nullable. Perhaps a warning would suffice for that?

IMO this would be enough to justify forcing a `default: null` line. It's short but gets rid of any and all ambiguity as to what's returned when no case is matched.

**Comment by Jetz72** [comment]
I do think it's good to avoid implicit nulls in situations where multiple meanings of null exist (which is why I kinda don't like that a nullable function can complete normally without an explicit return), but it can be handy when there's only one meaning (e.g. a map with a non-nullable value type returning null to signify the key isn't present reduces a separate `containsKey` check down to a quick inline `?.`). It'd be nice to, at minimum, allow an implicit `default: null` in situations where all other cases (or the surrounding context of the expression) are non-nullable.

In instances where you know the switch statement to be functionally exhaustive (even if static analysis can't be certain), it can also be a bit irritating to have to specify handling that shouldn't be able to happen, e.g.
```dart
int x = someValue();
if(x <= 0)
  return handleNegative();
else if(x <= 3) {
  var handlerText = switch(x) {
    case 1 => "one";
    case 2 => "two";
    case 3 => "three";
    default => throw AssertionError("someone invented a new number without telling me");
  };
  return handleSmallValue(handlerText);
}
else if(x <= 10) {
  …
```

**Comment by munificent** [comment]
Exhaustiveness checking is still a TODO in the proposal, but my intention is that the cases in a switch expression must be exhaustive or you get a compile error. You can make a switch expression trivially exhaustive by adding `default` or `case _`.

I'm a little less certain about how we should handle non-exhaustive switch *statements*. I'm inclined to make that a compile error too because a key part of the value of pattern matching is these exhaustiveness checks. At the same time, there's no soundness hole in having it non-exhaustive. Perhaps it should be a warning? I think that's how we handle non-exhaustive cases when switching on an enum today.

**Comment by fishythefish** [comment]
Continuing from email:

@eernstg wrote:
> Note that a switch expression can't just omit the default branch in a production mode compilation: the evaluation of an expression must yield an object (or terminate the evaluation abruptly, e.g., by throwing), it can't just "fall through".

No, I disagree with this. In production compilations, we routinely deviate from the spec in order to avoid unnecessary overhead, and yes, this may result in undefined behavior if an API is used out of contract. In the email thread, we weren't talking about eliminating *all* `default` cases; we were talking about using an `unreachable()` function as a compiler hint that the case is unreachable.

Suppose I know `n` is always `1`, `2`, or `3`, and I want to represent it as a string.

```dart
var s = switch(n) {
  case 1 => "one";
  case 2 => "two";
  case 3 => "three";
};
```

This snippet would be rejected statically since the cases are inexhaustive. I can add a default case to placate the compiler:

```dart
var s = switch(n) {
  case 1 => "one";
  case 2 => "two";
  case 3 => "three";
  default => unreachable();
};
```

In debug/spec mode, I'm fine with generating the additional code to support this case, but in production, I'm adamant that we'll never reach the `default`, so I don't want to pay the extra cost. In production mode, `dart2js` could recognize the `unreachable()` invocation, assume that it and everything it dominates is unreachable, and potentially generate smaller code.

If `n` turns out to be `4`, we make no promises as to the behavior. We assume that all logic has been tested adequately in debug mode/DDC. Depending on how this code is compiled, we may be able to recognize the issue at compile time, or we may still end up throwing at runtime, or we may produce a bogus value, etc.

**Comment by eernstg** [comment]
> If `n` turns out to be `4`, we make no promises as to the behavior.

Right, my thinking was too narrowly focused on the surface language constructs. A compiler could of course compile a switch expression to, for instance, a computation that stores the value of the expression somewhere, and then the missing default case would just not store anything (it wouldn't do anything at all), and the expression evaluation would then behave as if it had yielded whatever value that storage location had already. So I agree that the `unreachable` branch could actually be omitted.
